## Exercise: 

You are a data analyst for ASDA Groceries (Walmart's UK subsidiary, https://groceries.asda.com/), looking after the checkout customer experience. A new product manager has started at the company and she asks you to give a short report on the recent performance of the checkout and to give your recommendations on areas of opportunity/concern. You have pulled the attached data from the company’s analytics reporting tool. 

Background on ASDA Groceries' checkout experience: 

The company has a 4-step checkout, where every user has to go through each step in order to complete the checkout 

Step 1 – Before You Go – a page where the user is shown a number of product recommendations based on what they have either previous bought or complementary product to those already in their cart 

Step 2 – Book A Slot – a page where the user selects the time and day they would like their delivery. Availability and cost of slot here can vary. 

Step 3 – Select Payment – a page where existing users can select a credit card that was previous used or where new/existing users can add a new credit card to use again an order 

Step 4 – Order Summary – a page where users can see a summary of their order including order contents, booked slot, payment type and a breakdown of the final cost. This is where a customer places an order 

For ease of analysis – assume no same user orders more than once in the same day 

## Task: 

Run some analysis on the data set and present your top 5 findings to the new product manager.  Please see above User checkout experience. Assuming you can get all the data you want: 
 - What would you measure and how will you make decisions from there? 
 - What would be the top 3 metrics to define the success of the product?  
 - What would be the dashboards/report you would build to track those? 

In [1]:
import pandas as pd

In [2]:
df = pd.read_excel('International Data Analyst Take Home Excercise 2.xlsx',
                  sheet_name = 'Data')
df.head()

Date Customer Type     Delivery Type Device Type  \
0 2017-09-01      Existing  Store Collection    Desktop    
1 2017-09-01           New     Home Delivery    Desktop    
2 2017-09-01           New  Store Collection    Desktop    
3 2017-09-01           New     Home Delivery     Mobile    
4 2017-09-01      Existing  Store Collection     Mobile    

   Before You Go Unique Daily Users  Before You Go Cart Adds   \
0                              4910                      1929   
1                              4450                      1753   
2                               487                       148   
3                               398                       103   
4                              5159                      1750   

   Before You Go Total Revenue From Cart Adds  Book Slot Unique Daily Users  \
0                                        3737                          4839   
1                                        2470                          4166   
2                                         257                           478   
3                                         153                           343   
4                                        3356                          4785   

   Select Payment Unique Daily Users  Order Summary Unique Daily Users  \
0                               4342                              4245   
1                               3608                              3504   
2                                452                               435   
3                                272                               261   
4                               4242                              4059   

   Orders   Revenue  
0    4037  379289.0  
1    3450  311501.0  
2     399   36074.0  
3     250   17452.0  
4    3855  360265.0

In [3]:
df.columns

Index(['Date', 'Customer Type', 'Delivery Type', 'Device Type',
       'Before You Go Unique Daily Users', 'Before You Go Cart Adds ',
       'Before You Go Total Revenue From Cart Adds',
       'Book Slot Unique Daily Users', 'Select Payment Unique Daily Users',
       'Order Summary Unique Daily Users', 'Orders', 'Revenue'],
      dtype='object')

In [4]:
# Run some analysis on the data set and present your top 5 findings to the new product manager.  
# Please see above User checkout experience. Assuming you can get all the data you want: 
# What would you measure and how will you make decisions from there? 
# What would be the top 3 metrics to define the success of the product?  
# What would be the dashboards/report you would build to track those? 

ATTN: !pip install plotly

## Customer Type vs Revenue

In [5]:
df_customer = df.groupby(['Customer Type']).sum()
df_customer['percentage'] = (df_customer['Revenue']/df_customer['Revenue'].sum()) * 100
df_customer[['Revenue', 'percentage']]

Revenue  percentage
Customer Type                          
Existing       1.176188e+08   91.772794
New            1.054424e+07    8.227206

In [15]:
import plotly.express as px
fig = px.scatter(df, x='Revenue', y='Before You Go Total Revenue From Cart Adds', color='Customer Type',
                 size='Revenue', marginal_y="violin", marginal_x="box")
fig.show()

- Existing customers are making most of the revenue, which is 91.77% of total revenue made in September. New customers are making 8.2% of revenues.
- The highest revenue made by the new customers is 311.8k, however the lowest revenue made by 274.6K. Indicating that the new customers can offer more revenues once they get used to the website.

## Delivery Type vs Revenue

In [9]:
df_delivery = df.groupby(['Delivery Type']).sum()
df_delivery['Percentage'] = (df_delivery['Revenue']/df_delivery['Revenue'].sum()) * 100
df_delivery[['Revenue', 'Percentage']]

Revenue  Percentage
Delivery Type                             
Home Delivery     1.059129e+08   82.639174
Store Collection  2.225016e+07   17.360826

In [12]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

labels = df['Delivery Type']
value_exist = df['Revenue'].loc[df['Customer Type'] == 'Existing']
value_new = df['Revenue'].loc[df['Customer Type'] == 'New']
colors = ['gold', 'darkorange']

fig = make_subplots(rows=1, cols=2, specs=[[{'type':'domain'}, {'type':'domain'}]])
fig.add_trace(go.Pie(labels=labels, values=value_exist, name="Existing Customer"),
              1, 1)
fig.add_trace(go.Pie(labels=labels, values=value_new, name="New Customer"),
              1, 2)

fig.update_traces(hole=.4, hoverinfo="label+percent+name", textinfo = 'value',
                 marker=dict(colors=colors, line=dict(color='#000000', width=2)))

fig.update_layout(
    title_text="Customer Type & Delivery Type",
    annotations=[dict(text='Existing', x=0.18, y=0.5, font_size=20, showarrow=False),
                 dict(text='New', x=0.82, y=0.5, font_size=20, showarrow=False)])
fig.show()

- Based on dataframe created above, customers would like to pick up in-store rather than selecting home delivery. And sore collection related revenue is more than twice the revenue made with home delivery.

- Based on Graphics comparing the revenue earned by existing and new customers and the delivery methods, the following information can be obtained:
- for existing customers, there are not huge delivery methods preferences, which means 44.7% of existing customers prefer store collecting shopping items and 55.3% of existing customers prefer home delivery.
- for new customers, they prefer store collecting than home delivery. only 37.7% of new customers prefer home delivery, but 62.3% of new customers prefer store collection.

## Device Type vs Revenue

In [29]:
df_device = df.groupby(['Device Type','Delivery Type']).sum()
df_device['Percentage'] = (df_device['Revenue']/df_device['Revenue'].sum()) * 100
df_device[['Revenue', 'Percentage']]

Revenue  Percentage
Device Type Delivery Type                            
Desktop     Home Delivery     58668333.32   45.776325
            Store Collection  11931744.02    9.309816
Mobile      Home Delivery     47244550.18   36.862849
            Store Collection  10318420.02    8.051010

In [14]:
fig_device = px.scatter_matrix(df, dimensions=['Orders', 'Revenue'], color='Device Type')
fig_device.show()

- 55% of revenues are made with Desktop and 45% of revenes are made with mobile.
- store collection type of delivery method is taking almost 1/6 of the total revenue when people order through desktop, and 1/5 of the revenue made when people order through mobiles.

## Date vs Revenue

In [31]:
df_date = df.groupby(['Date'],as_index=False).sum()
df_date['Percentage'] = (df_date['Revenue']/df_date['Revenue'].sum()) * 100
df_date[['Date', 'Percentage']]

Date  Percentage
0  2017-09-01    3.408493
1  2017-09-02    3.463405
2  2017-09-03    3.409490
3  2017-09-04    3.437120
4  2017-09-05    3.483982
5  2017-09-06    3.099426
6  2017-09-07    3.363887
7  2017-09-08    3.366551
8  2017-09-09    3.341472
9  2017-09-10    3.419284
10 2017-09-11    3.415229
11 2017-09-12    3.419688
12 2017-09-13    3.461051
13 2017-09-14    3.481053
14 2017-09-15    3.485640
15 2017-09-16    3.468109
16 2017-09-17    3.367450
17 2017-09-18    3.399300
18 2017-09-19    3.352345
19 2017-09-20    3.364454
20 2017-09-21    3.198685
21 2017-09-22    3.228946
22 2017-09-23    3.259228
23 2017-09-24    3.153355
24 2017-09-25    3.186718
25 2017-09-26    3.128052
26 2017-09-27    3.177188
27 2017-09-28    3.153113
28 2017-09-29    3.256613
29 2017-09-30    3.250670

In [38]:
df_date['Revenue'].nlargest(5)

14    4467303.0
4     4465178.0
13    4461424.0
15    4444834.0
1     4438806.0
Name: Revenue, dtype: float64

In [35]:
fig_d = px.scatter(df_date, x='Date', y='Revenue', color='Date',
                 color_continuous_scale=px.colors.sequential.Cividis_r)

fig_d.show()

- September 6th made the minimum revenue in September. 
- September 15th made the maximum revenue in September. 
- The differences are not very large.
- 5 top revenue made by daily are: 9/15, 9/5, 9/14, 9/16, 9/2, which are the first two weeks, probably because of labor day.

## Before You Go Total Revenue vs Revenue

In [45]:
df_tr = df_tr.groupby(['Date'], as_index=False).sum()
df_tr = df_tr[['Date','Before You Go Total Revenue From Cart Adds', 'Revenue']]
df_tr['percentage'] = (df_tr['Before You Go Total Revenue From Cart Adds']/df_tr['Revenue']) * 100

In [51]:
df_tr

Date  Before You Go Total Revenue From Cart Adds     Revenue  \
0  2017-09-01                                       38381  4368428.00   
1  2017-09-02                                       37953  4438806.00   
2  2017-09-03                                       38345  4369706.00   
3  2017-09-04                                       37871  4405118.00   
4  2017-09-05                                       37629  4465178.00   
5  2017-09-06                                       37569  3972319.00   
6  2017-09-07                                       38926  4311260.00   
7  2017-09-08                                       38164  4314674.00   
8  2017-09-09                                       38184  4282532.00   
9  2017-09-10                                       37658  4382259.00   
10 2017-09-11                                       37781  4377062.00   
11 2017-09-12                                       38175  4382776.00   
12 2017-09-13                                       38383  4435789.00   
13 2017-09-14                                       37675  4461424.00   
14 2017-09-15                                       37617  4467303.00   
15 2017-09-16                                       37589  4444834.00   
16 2017-09-17                                       38232  4315827.00   
17 2017-09-18                                       37794  4356647.00   
18 2017-09-19                                       38211  4296468.00   
19 2017-09-20                                       37703  4311987.00   
20 2017-09-21                                       42643  4099532.70   
21 2017-09-22                                       41603  4138315.22   
22 2017-09-23                                       43325  4177125.94   
23 2017-09-24                                       42257  4041436.00   
24 2017-09-25                                       42912  4084194.72   
25 2017-09-26                                       42912  4009006.94   
26 2017-09-27                                       42453  4071980.36   
27 2017-09-28                                       43371  4041125.80   
28 2017-09-29                                       42325  4173774.84   
29 2017-09-30                                       42532  4166158.02   

    percentage  
0     0.878600  
1     0.855027  
2     0.877519  
3     0.859705  
4     0.842721  
5     0.945770  
6     0.902891  
7     0.884516  
8     0.891622  
9     0.859328  
10    0.863159  
11    0.871023  
12    0.865303  
13    0.844461  
14    0.842052  
15    0.845678  
16    0.885856  
17    0.867502  
18    0.889358  
19    0.874376  
20    1.040192  
21    1.005312  
22    1.037196  
23    1.045594  
24    1.050684  
25    1.070390  
26    1.042564  
27    1.073241  
28    1.014070  
29    1.020893

In [48]:
percentage = (df_tr['Before You Go Total Revenue From Cart Adds'].sum()/df_tr['Revenue'].sum()) * 100
percentage

0.9255187222586858

In [49]:
df[['Before You Go Unique Daily Users', 'Before You Go Cart Adds ',
            'Before You Go Total Revenue From Cart Adds']].sum()

Before You Go Unique Daily Users              1680042
Before You Go Cart Adds                        649191
Before You Go Total Revenue From Cart Adds    1186173
dtype: int64

In [50]:
df[['Before You Go Unique Daily Users', 'Order Summary Unique Daily Users']]

Before You Go Unique Daily Users  Order Summary Unique Daily Users
0                                4910                              4245
1                                4450                              3504
2                                 487                               435
3                                 398                               261
4                                5159                              4059
..                                ...                               ...
235                               411                               243
236                              4800                              3781
237                              5029                              3582
238                             20698                             15919
239                                61                                40

[240 rows x 2 columns]

- the revenue made through adds are around 1% of the revenue made calculated based on every single day in September. So revenue from adds are still taking the minimum percentile.

#### # What would you measure and how will you make decisions from there? 
I will measure elements like 
'Before You Go Unique Daily Users': to get some idea how many customers are getting into the website; 

'Before You Go Cart Adds': to understand how many adds are posted

'Before You Go Total Revenue From Cart Adds': how efficiently the adds can work

'Order Summary Unique Daily Users': if the most of customers are making orders, or are they not happy with the add

#### # What would be the top 3 metrics to define the success of the product?  
1. if the revenue are some how increasing

2. if the customers are increasing because we can offer better/related adds

3. what proportion does adds revenue take in the total revenue

#### # What would be the dashboards/report you would build to track those? 

first of all, I need to calculate the total revenue to see if the total revenue is increasing or keeping in the same level. And then, calculate the proportion of cart adds revenue to the total revenue to see if adds works as what we suppose to be. If not, find if the number of total unique users to the total adds and order summary to see if adds are influencing the purchase. And finally, get the date and revenue relations.